Update README (#201)
probe-rs/probe-rs: README.md

# probe-rs

a modern, embedded debugging toolkit,
written in Rust

[![crates.io](https://meritbadge.herokuapp.com/probe-rs)](https://crates.io/crates/probe-rs) [![documentation](https://docs.rs/probe-rs/badge.svg)](https://docs.rs/probe-rs) [![Actions Status](https://github.com/probe-rs/probe-rs/workflows/CI/badge.svg)](https://github.com/probe-rs/probe-rs/actions) [![chat](https://img.shields.io/badge/chat-probe--rs%3Amatrix.org-brightgreen)](https://matrix.to/#/#probe-rs:matrix.org)

The goal of this library is to provide a toolset to interact with a variety of embedded MCUs and debug probes.

Similar projects like OpenOCD, PyOCD, Segger Toolset, ST Tooling, etc. exist.

> They all implement the GDB protocol and their own protocol on top of it to enable GDB to communicate with the debug probe.

Reviewer: This is not really true for the Segger tooling, they have their own DLL which you can use.

> This project gets rid of the GDB layer (even tho we provide one you can use until [Microsoft DAP](https://code.visualstudio.com/blogs/2018/08/07/debug-adapter-protocol-website) support is fully implemented) and provides a direct interface to the debug probe,

Reviewer: I would remove the sentence in brackets, and maybe add a separate section for the `gdb-server` and `vscode` projects?

which then enables other software to use it's debug functionality.

**The end goal of this project is to have a complete library toolset to enable other tools to communicate with embedded targets.**

## Functionality

> As of 27.03.2020 this lib can

Reviewer: I would change this to the version, not the date.

# probe-rs

a modern, embedded debugging toolkit,
written in Rust

[![crates.io](https://meritbadge.herokuapp.com/probe-rs)](https://crates.io/crates/probe-rs) [![documentation](https://docs.rs/probe-rs/badge.svg)](https://docs.rs/probe-rs) [![Actions Status](https://github.com/probe-rs/probe-rs/workflows/CI/badge.svg)](https://github.com/probe-rs/probe-rs/actions) [![chat](https://img.shields.io/badge/chat-probe--rs%3Amatrix.org-brightgreen)](https://matrix.to/#/#probe-rs:matrix.org)

The goal of this library is to provide a toolset to interact with a variety of embedded MCUs and debug probes.

Similar projects like OpenOCD, PyOCD, Segger Toolset, ST Tooling, etc. exist.
They all implement the GDB protocol and their own protocol on top of it to enable GDB to communicate with the debug probe.
Only Segger provides a closed source DLL which you can use for talking to the JLink.

This project gets rid of the GDB layer and provides a direct interface to the debug probe,
which then enables other software to use it's debug functionality.

**The end goal of this project is to have a complete library toolset to enable other tools to communicate with embedded targets.**

## Functionality

> As of 0.6.0 this lib can

Reviewer: ```suggestion
As of version 0.6.0 this library can
```

- connect to a DAPLink, STLink or JLink
- talk to ARM and Risc-V cores via SWD or JTAG
- read and write arbitrary memory of the target
- halt, run, step, breakpoint and much more the core
- download ELF, BIN and IHEX binaries using standard CMSIS-Pack flash algorithms to ARM cores

> - provide debug information about the target state (stacktrace, stackframe, etc)

Reviewer: ```suggestion
- provide debug information about the target state (stacktrace, stackframe, etc.)
```

To see what new functionality was added have a look at the [CHANGELOG](CHANGELOG.md)

## Downloading a file

The `cargo-flash` utility can be used as a cargo subcommand to download a compiled Rust program onto a target device. It can also be used to download arbitrary ELF files that might come out of a C/C++ compiler. Have a look at [cargo-flash](https://github.com/probe-rs/cargo-flash) for more information.

## GDB

We provide a GDB stub you can use until [Microsoft DAP](https://code.visualstudio.com/blogs/2018/08/07/debug-adapter-protocol-website) support is fully implemented.
You can find it [here](https://github.com/probe-rs/probe-rs/tree/master/gdb-server) and you can also use it from within `cargo-flash` with the `--gdb` flag.

## VScode

> We are implementing [Microsoft DAP](https://code.visualstudio.com/blogs/2018/08/07/debug-adapter-protocol-website) to provide full probe-rs integration into modern debuggers such as the built in one of VSCode.

Reviewer: If we mention this here, I wouldn't mention it in the GDB section. Also, the debugger is not really in VS Code, the main part will be in probe-rs. I would also write VS Code the same everywhere, in the section heading and here.

Author: This is not in the GDB section, there is a VSCode section inbetween ;)

Reviewer: This is the VS Code section, yes, but Microsoft DAP is also mentioned in the GDB section.

Author: Ah, whoops, sorry, didn't understand that properly ...

### I need help!

Don't hesitate to [file an issue](https://github.com/probe-rs/probe-rs/issues/new), ask questions on [Matrix](https://matrix.to/#/#probe-rs:matrix.org), or contact [@Yatekii](https://github.com/Yatekii) via e-mail.

### How can I help?

Please have a look at the issues or open one if you feel that something is needed.

> Any contibutions are very welcome!

Reviewer: ```suggestion
Any contributions are very welcome!
```

> Any code contibutions are very welcome!

Reviewer: I think other contributions (e.g. documentation or testing) are welcome as well, so I would not mention code here.
```suggestion
Any contributions are very welcome!
```

 * MIT license ([LICENSE-MIT](LICENSE-MIT) or
   http://opensource.org/licenses/MIT) at your option.

> 

Reviewer: I would not completely remove this, maybe just change the wording. I think pyOCD was pretty essential for the initial flash code, so I would keep an acknowledgement section.
